### ISSUE TITLE
NFS Ganesha, file transfer success despite the lack of space

### ISSUE THREAD

**Issue by ppp368** [leil-io/saunafs#233]
If there is less space than the size of the transferred file (or no space at all), then the transfer stops for several seconds.
Then continues and ends "successfully", without any error or log. When there is no more free space file is truncated.

## Software versions
OS: `Ubuntu 22.04LTS`

SaunaFS and Ganesha Version: `4.5.1-20240928-051207-stable-main-3aa2a609` (both from SaunaFS repositories)

## Setup
For easy reproduction, steps are performed on a single machine with SaunaFS Master, Chunkserver and NFS Ganesha.

Configuration files:

* `/etc/saunafs/sfsmaster.cfg`

    ```cfg
    PERSONALITY = master
    MASTER_HOST = sfsmaster
    MASTER_PORT = 9419
    MASTER_TIMEOUT = 10
    AUTO_RECOVERY = 1
    REJECT_OLD_CLIENTS = 1
    AVOID_SAME_IP_CHUNKSERVERS = 1
    MATOCS_LISTEN_HOST = *
    MATOCS_LISTEN_PORT = 9420
    MATOCL_LISTEN_HOST = *
    MATOCL_LISTEN_PORT = 9421
    OPERATIONS_DELAY_DISCONNECT = 300
    ```

* `/etc/saunafs/sfsexports.cfg`

    ```cfg
    *                       /       rw,alldirs,maproot=0,ignoregid
    *                       .       rw
    ```

* `/etc/saunafs/sfsgoals.cfg`

    ```cfg
    1 1 : _
    2 2 : _ _
    3 3 : _ _ _
    4 4 : _ _ _ _
    5 5 : _ _ _ _ _
    10 ec21 : $ec(2,1)
    11 ec31 : $ec(3,1)
    12 ec32 : $ec(3,2)
    ```

* `/etc/ganesha/ganesha.conf`

    ```conf
    NFS_KRB5 {
        Active_krb5 = false;
    }
    NFSV4 {
        Grace_Period = 5;
        Lease_Lifetime = 10;
    }
    EXPORT {
        Export_Id = 1;
        Path = "/";
        Pseudo = "/";
        Access_Type = RW;
        Squash = none;
        Protocols = 3, 4;
        Attr_Expiration_Time = 0;

        FSAL {
            Name = SaunaFS;
            hostname = "10.244.1.13";
            port = "9421";
            password = "*****";
        }
    }
    ```

## Reproduction steps

```sh
# Mount using NFS
$ mkdir -p /mnt/nfs
$ mount -vvvv <NFS SERVER IP>:/ /mnt/nfs

# Create a file larger than the available space and upload it
$ head -c <SIZE> /dev/random > test_file
$ pv test_file > /mnt/nfs/test_file
```

**Comment by ralcolea** [comment]
Hi @ppp368,

I have been working in this issue in the last days and I share my conclusions below:
- I added new tests to verify how the NFS client handles ENOSPC (no space left on device) errors in similar scenarios and all the tests passed.

- The only change I did was in the error handler to translate correctly `SAUNAFS_ERROR_NOSPACE` (SaunaFS error) to `ENOSPC` (POSIX error). It did not change the behavior of reporting the error, but the error type reported by NFS client.

- The line `pv test_file > /mnt/nfs/test_file` fails to report errors because of caching. With standard I/O (default), pv may not notice write errors (like "no space left on device") until much later.

- This behavior is especially noticeable because:
   - NFS often buffers writes on the client side before sending them to the server.
   - The OS page cache and NFS write buffering can hide write errors from applications.

- The solution I used to detect write errors inmediately when using pv was to use `--direct-io`:
`pv --direct-io test_file > /mnt/nfs/test_file`

- Using `--direct-io` ensures that pv bypasses these caches and interacts directly with the NFS server, catching errors more reliably.
